File: 이민희/DP/boj1463.cpp
#include <iostream>
#include <vector>
using namespace std;

int main(){
    int N;
    
    cin >> N;
    vector<int> dp(N + 1);

    dp[0] = 0;
    dp[1] = 0;
    dp[2] = 1;
    dp[3] = 1;

    for (int i = 4; i <= N; i++)
    {
        dp[i] = dp[i - 1] + 1; // 방법 1
        if (i % 2 == 0) dp[i] = min(dp[i], dp[i / 2] + 1); // 방법 1과 방법 2를 비교하여 둘 중 작은 걸 선택
        if (i % 3 == 0) dp[i] = min(dp[i], dp[i / 3] + 1); // 방법 1과 방법 2를 비교하여 둘 중 작은 걸 선택
    }

    cout << dp[N];
}

File: 이민희/DP/boj1463.py
n = int(input())

dp = [0 for _ in range(n + 100)]
dp[1] = 0
dp[2] = 1
dp[3] = 1

if (n < 4) :
    print(dp[n])
else :
    for i in range(4, n + 1, 1) :
        num = dp[i - 1] + 1
        if (i % 3 == 0) :
            num = min(dp[i // 3] + 1, num)
        if (i % 2 == 0) :
            num = min(dp[i // 2] + 1, num)
        dp[i] = num

    print(dp[n])